server/src/updaters/relationship-updaters.js

```diff
@@ -0,0 +1,202 @@
+// @flow
+
+import type { Viewer } from '../session/viewer';
+
+import invariant from 'invariant';
+
+import {
+  type RelationshipStatus,
+  relationshipStatuses,
+  type RelationshipRequest,
+} from 'lib/types/relationship-types';
+import { ServerError } from 'lib/utils/errors';
+
+import { dbQuery, SQL } from '../database';
+
+async function updateRelationship(
+  viewer: Viewer,
+  request: RelationshipRequest,
+) {
+  const { userID, status } = request;
+
+  if (viewer.id === userID) {
+    throw new ServerError('invalid_user');
+  }
+
+  const userQuery = SQL`SELECT 1 FROM users WHERE id = ${userID} LIMIT 1`;
+  const [users] = await dbQuery(userQuery);
+
+  if (!users.length) {
+    throw new ServerError('user_does_not_exist');
+  }
+
+  // friend request
+  if (status === relationshipStatuses.PENDING_FRIEND) {
+    const [relationships] = await fetchRelationshipQuery(viewer.id, userID);
+
+    if (!relationships.length) {
+      await createRelationshipQuery(viewer.id, userID, status);
+    } else {
+      const [actorRelationship] = getRelationshipsForActorUser(
+        relationships,
+        viewer.id,
+      );
+      if (actorRelationship) {
+        const message = actorFriendshipErrorMap[actorRelationship.status];
```

Comment:
I don't think we always need to fail here. `invitation_already_sent_by_requester` in particular should just return a success in my opinion - it's better for operations to be idempotent whenever possible